grpc_server/lib/grpc/server/adapters/thousand_island.ex

```diff
@@ -0,0 +1,639 @@
+    │                         │                          │<──── pull next ───────┤
+    │                         │                          │       (req2)          │
+    │                         │                          ├─ process req2         │
+    │                         │                          ├─ send_reply(resp2)    │
+    │                         │<── :grpc_send_data ──────┤                       │
+    │<─── DATA(resp2) ────────┤                          │                       │
+    ├─── DATA (END_STREAM) ──>│                          │                       │
+    │                         ├─ finish stream ──────────┼──────────────────────>│
+    │                         │                          │<──── nil (done) ──────┤
+    │                         │                          ├─ handler finishes     │
+    │                         │                          ├─ send_trailers        │
+    │                         │<── :grpc_send_trailers ──┤                       │
+    │<─── HEADERS(trailers) ──┤                          │                       │
+    │    (END_STREAM)         │                          x                       x
+  ```
+
+  ## Key Design Patterns
+
+  1. **Async Message Passing**: User handlers send messages to Handler process
+     - Decouples user code from HTTP/2 frame management
+     - Handler serializes all socket writes (thread-safe)
+
+  2. **Lazy Enumerables**: Streaming requests use `Stream.unfold`
+     - Backpressure: handler blocks if no messages available
+     - Memory efficient: doesn't buffer entire stream
+
+  3. **Header Accumulation**: Inspired by Cowboy's `set_resp_headers`
+     - Accumulate headers in Handler state (not process dictionary!)
+     - Send on first DATA or trailers (whichever comes first)
+     - Allows custom headers while respecting HTTP/2 constraints
+
+  4. **GenServer Message Queue**: BidiStream acts as message buffer
+     - Decouples incoming frame rate from handler processing rate
+     - Natural backpressure via GenServer mailbox
+
+  5. **Process Dictionary for Stream Metadata**: 
+     - Used in Dispatcher context: `Process.put({:bidi_stream_pid, stream_id}, pid)`
+     - Allows Connection to find BidiStream when DATA arrives
+     - Alternative to passing state through deep call stack
+  """
+
+  @behaviour GRPC.Server.Adapter
+
+  require Logger
+  alias GRPC.Server.Adapters.ThousandIsland.Handler
+
+  @default_num_acceptors 100
+  @default_max_connections 16384
+
+  @doc """
+  Starts a ThousandIsland server.
+
+  ## Options
+    * `:ip` - The IP to bind the server to (default: listen on all interfaces)
+    * `:port` - The port to listen on (required)
+    * `:num_acceptors` - Number of acceptor processes (default: 100)
+    * `:num_connections` - Maximum concurrent connections (default: 16384)
+    * `:transport_options` - Additional transport options to pass to ThousandIsland
+  """
+  @impl true
+  def start(endpoint, servers, port, opts) do
```

Comment:
this should be start_link

it's a bit misleading a start function that links a process (or 2 in the current version)